daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala

```diff
@@ -16,6 +16,49 @@ import scala.annotation.tailrec
 
 private[validation] object Typing {
 
+  def typeOf(env: Env, exp: Expr): Type = { // testing entry point
+    // this implementation is stack-safe
+    runWork(env.typeOfExpr(exp))
+  }
+
+  // stack-safety achieved via a Work trampoline.
+  private sealed abstract class Work[A]
```

Comment:
I added `flapMap` as a shim around `Bind`
The main benefit is it avoid the need for the `Any` hack. See comments below,

Reply:
Reality is that `Any` is still here (even though it's no longer immediately obvious). The key place to observe this is `runWork`:
- `case Ret(x) =>` means `x` has type `Any` (as the real type has been erased)
- `work1.flatMap(x => k1(x).flatMap(k))` mean `x` also has type `Any` (again due to type erasure).

The key risk here is that we may introduce runtime exceptions due to casts not being possible. 

That said, the 2 usage sites for this method are:
- `typeOf` which is `validation` package private
- and `checkTopExpr` which is private.

So it may well be that we don't actually need to care about type erasure in these cases?

Reply:
I prefer the `Any` to be explicit, so I inlined/removed the `flatMap`

The code is now:
```
          case Bind(work1, k1) => Bind(work1, ((x: Any) => Bind(k1(x), k)))
```

In reality `x` has existential type `X`, because of the definition:
```
    final case class Bind[A, X](work: Work[X], k: X => Work[A]) extends Work[A]
```

And so the one safe thing we can do with `x` is pass it as an argument to `k1`

Reply:
Maybe there is a better scala idiom to make this clear. :shrug:

Reply:
In Scala, `TypeTag` is normally how you manage type erasure in pattern matching. That said, it can sometimes get messy using this (certainly my naive code experiments didn't quickly resolve issues here)!